ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+		default:
+		}
+	}
+
+	// Don't wait for checkpointing to finish because it might take too long.
+}
+
+// processUpdateResult sends WAL update result using ResultCh channel
+// and waits for signal from DoneCh channel.
+// This ensures that WAL update and ledger state update are in sync.
+func processUpdateResult(update *WALTrieUpdate, updateResult error) {
+	// Send result of WAL update
+	update.ResultCh <- updateResult
+
+	// Wait for trie update to complete
+	<-update.DoneCh
+}
+
+func (c *Compactor) checkpoint(ctx context.Context, tries []*trie.MTrie, checkpointNum int) error {
+
+	err := createCheckpoint(c.checkpointer, c.logger, tries, checkpointNum)
+	if err != nil {
+		return fmt.Errorf("cannot create checkpoints: %w", err)
+	}
+
+	// Return if context is canceled.
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+	}
+
+	err = cleanupCheckpoints(c.checkpointer, int(c.checkpointsToKeep))
+	if err != nil {
+		return fmt.Errorf("cannot cleanup checkpoints: %w", err)
+	}
+
+	if checkpointNum > 0 {
+		for observer := range c.observers {
+			// Don't notify observer if context is canceled.
+			// observer.OnComplete() is called when Compactor starts shutting down,
+			// which may close channel that observer.OnNext() uses to send data.
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+				observer.OnNext(checkpointNum)
+			}
+		}
+	}
+
+	return nil
+}
+
+func createCheckpoint(checkpointer *realWAL.Checkpointer, logger zerolog.Logger, tries []*trie.MTrie, checkpointNum int) error {
+
+	logger.Info().Msgf("serializing checkpoint %d", checkpointNum)
+
+	startTime := time.Now()
+
+	writer, err := checkpointer.CheckpointWriter(checkpointNum)
+	if err != nil {
+		return fmt.Errorf("cannot generate checkpoint writer: %w", err)
+	}
+	defer func() {
+		closeErr := writer.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	err = realWAL.StoreCheckpoint(writer, tries...)
+	if err != nil {
+		return fmt.Errorf("error serializing checkpoint (%d): %w", checkpointNum, err)
+	}
+
+	duration := time.Since(startTime)
+	logger.Info().Float64("total_time_s", duration.Seconds()).Msgf("created checkpoint %d", checkpointNum)
+
+	return nil
+}
+
+func cleanupCheckpoints(checkpointer *realWAL.Checkpointer, checkpointsToKeep int) error {
+	// Don't list checkpoints if we keep them all
+	if checkpointsToKeep == 0 {
+		return nil
+	}
+	checkpoints, err := checkpointer.Checkpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(checkpointsToKeep) {
+		// if condition guarantees this never fails
+		checkpointsToRemove := checkpoints[:len(checkpoints)-int(checkpointsToKeep)]
+
+		for _, checkpoint := range checkpointsToRemove {
+			err := checkpointer.RemoveCheckpoint(checkpoint)
+			if err != nil {
+				return fmt.Errorf("cannot remove checkpoint %d: %w", checkpoint, err)
+			}
+		}
+	}
```

Comment:
better to split into two functions, one to compute what files to remove, the other to remove by the specified files.

The first function would be easier to write unittest for.